### DISCUSSION TITLE
SGP4 vs SDP4

### DISCUSSION THREAD

**Discussion by FIU-ELF** [brandon-rhodes/pyephem#195] [Q&A]
Hi Brandon,

I noticed PyEphem has both SGP4 and SDP4 code in it. Does it select the best propagator for the problem or do we have to somehow select it in Python?

Thank you,
Chris

**Comment by brandon-rhodes** [comment]
It’s been a long time since I’ve looked at that part of `libastro`, so let’s do a quick search of the repository using the GitHub search bar at the top of the page:

https://github.com/search?q=sdp4+repo%3Abrandon-rhodes%2Fpyephem&type=Code&ref=advsearch&l=&l=

The only call that I see to the routine in the codebase is here:

https://github.com/brandon-rhodes/pyephem/blob/1bb605c6c6fd99b55521953e02229e3fa43baa88/libastro-3.7.7/earthsat.c#L388

It looks like it chooses based on the satellite's rate of revolution, without input from the user:

```
	if (se.se_XNO >= (1.0/225.0))
	    sgp4(&sd, &posvec, &velvec, dt); /* NEO */
	else
	    sdp4(&sd, &posvec, &velvec, dt); /* GEO */
```

I'm basing my interpretation of `se_XNO` a few lines up:

```
	se.se_XNO = op->es_n * (2*PI/MPD);	/* revs/day to rads/min */
```

Unless there's a way for users to manually change `es_n` without materially changing the satellite’s other attributes (and someone would have to do a deeper dive into the code to see if that’s possible or not), it looks like both routines can be called, and are selected based on one of the satellite’s own properties rather than a user option.

**Reply by FIU-ELF** [reply]
thank you for the explanation and for teaching me a search function I didn't know about! That is perfect (and astronomically accurate in terms of revs/day)! I really appreciate it.

**Reply by brandon-rhodes** [reply]
> thank you for the explanation and for teaching me a search function I didn't know about!

I'm very glad it came across as educational (which was my goal) and not “you could have done this yourself”!

From the comments, I suppose `NEO` and `GEO` must be “near earth orbit” and “geosynchronous orbit”.
